File: files/CH-179.R
library(tidyverse)
library(readxl)

path = "files/CH-179 Reshape a table.xlsx"
input = read_excel(path, range = "B2:F10") %>% as.matrix()
test  = read_excel(path, range = "H2:J14")

result = input %>%
  t() %>%
  as.vector() %>%
  na.omit() %>%
  matrix(ncol = 3, byrow = TRUE) %>%
  as.data.frame() %>%
  setNames(c("Date", "Product ID", "Quantity")) %>%
  mutate(Date = as.POSIXct(janitor::excel_numeric_to_date(as.numeric(Date))),
         Quantity = as.numeric(Quantity))

all.equal(result, test, check.attributes = F)
#> [1] TRUE

File: files/CH-179.py
import pandas as pd
import numpy as np

path = "CH-179 Reshape a table.xlsx"
input = pd.read_excel(path, usecols="B:F", skiprows=1, nrows=8).to_numpy()
test = pd.read_excel(path,  usecols="H:J", skiprows=1, nrows=13)

result_df = pd.DataFrame(input.flatten()[~pd.isna(input.flatten())].reshape(-1, 3), columns=["Date", "Product ID", "Quantity"])
result_df["Quantity"] = result_df["Quantity"].astype(np.int64)

print(result_df.equals(test)) # True
